### DISCUSSION TITLE
Multiple deployments for code reviews and testing

### DISCUSSION THREAD

**Discussion by objecttothis** [opensourcepos/opensourcepos#3671] [Team Posts]
@jekkos @daN4cat Is it possible for Pull requests to get their own build deployment demos for testing and peer code reviews?  Currently, when a commit gets pushed to a PR, Travis-CI builds it and pushes it out to the demo where it can be played with.  This is a little problematic though:
- Ideally the demo would be the latest stable release for perspective new users to play around with before installing on their own boxes so putting a dev version potentially causes users to think the software is less stable than it is.
- any time another dev pushes a commit, the build gets overwritten which means its more difficult for anyone peer reviewing the PR.

Is it possible to instead keep demo.opensourcepos.org to a release version and have Travis-CI build a PR triggered by a commit but instead in opensourcepos.og/[branch_name]/?  then the build maybe could be automatically wiped when either the branch gets deleted or the PR gets merged?  I know nothing of the automation but see the need. it would save lots of time in manually deploying builds to do a peer review.

**Comment by jekkos** [comment]
@objecttothis Good that you raise this, I was thinking about the exact same thing today. I have currently automated demo to restart and wipe the instance every hour, as it is heavily used and customized in  driveby fashion.

This scenario is certainly not compatible with what we desire as a situation for development and testing. As you say, it would be better to have a dedicated env in which we can easily see what is installed and override it if desired. 

So that brings me back to my initial idea to try and deploy ospos to a (managed) k8s cluster for demo and dev purposes and indeed separate one instance for internal validation and one for public usage. I'll add it to my todo list.

**Comment by objecttothis** [comment]
@jekkos, what hardware is needed for this?  If it's just for dev purposes, I can donate a FreeBSD jail on my server, but I'm unfamiliar with setting up and maintaining the cluster, so I would just need to give someone SSH access and let them set it up and maintain it.

**Comment by objecttothis** [comment]
Here's an article on deploying on FreeBSD.  https://www.freebsdnews.com/2020/02/28/how-to-deploy-kubernetes-cluster-on-freebsd-bhyve-by-cbsd/

**Comment by jekkos** [comment]
I was thinking to run it on DO. In fact I think you need three nodes minimum for k8s, so it could be at least 20 or 30 dollars a month. If I set it up I'd like to make it more or less self sustainable in terms of costs/

But for the short run maybe it's better to see if I can to adapt the deploy script to spinup a second container next to the demo on my VPS.

**Comment by jekkos** [comment]
Well I'm not very familiar with BSD, but I guess that the philosophy behind it is appealing. I'll have a look

**Comment by objecttothis** [comment]
No pressure.  If I can help, I'd be glad to... it would definitely streamline peer review and testing for everyone afterall to have something like this.
